packages/cli/src/scaling/scaling.service.ts

```diff
@@ -230,4 +257,86 @@ export class ScalingService {
 
 		throw new ApplicationError('This method must be called on a `worker` instance');
 	}
+
+	// #region Queue recovery
+
+	private readonly queueRecoverySettings: QueueRecoverySettings = {
+		batchSize: config.getEnv('executions.queueRecovery.batchSize'),
+		waitMs: config.getEnv('executions.queueRecovery.interval') * 60 * 1000,
+	};
+
+	scheduleQueueRecovery(waitMs = this.queueRecoverySettings.waitMs) {
+		this.queueRecoverySettings.timeout = setTimeout(async () => {
+			try {
+				const nextWaitMs = await this.recoverFromQueue();
+				this.scheduleQueueRecovery(nextWaitMs);
+			} catch (error) {
+				this.logger.error('[ScalingService] Failed to recover dangling executions from queue', {
+					msg: this.toErrorMsg(error),
+				});
+				this.logger.error('[ScalingService] Retrying...');
+
+				this.scheduleQueueRecovery();
+			}
+		}, waitMs);
+
+		const wait = [this.queueRecoverySettings.waitMs / (60 * 1000), 'min'].join(' ');
+
+		this.logger.debug(`[ScalingService] Scheduled queue recovery check for next ${wait}`);
+	}
+
+	stopQueueRecovery() {
+		clearTimeout(this.queueRecoverySettings.timeout);
+	}
+
+	/**
+	 * Mark in-progress executions as `crashed` if stored in DB as `new` or `running`
+	 * but absent from the queue. Return time until next recovery cycle.
+	 */
+	private async recoverFromQueue() {
+		const { waitMs, batchSize } = this.queueRecoverySettings;
+
+		const storedIds = await this.executionRepository.getInProgressExecutionIds(batchSize);
+
+		if (storedIds.length === 0) {
+			this.logger.debug('[ScalingService] Completed queue recovery check, no dangling executions');
+			return waitMs;
+		}
```

Comment:
`getInProgressExecutionIds` seems to get executions with status `new` and `running`, but not `waiting`. If there are only `waiting` executions in DB, they are not checked as we short-circuit here. Is that intentional?

Reply:
Queue recovery reconciles Redis state with DB state, so if Redis knows an execution not to be in any state in the Bull queue anymore, then the DB status is stale and needs to be overridden to `crashed`. 

In your example, if a DB has zero `new` or `running` executions, e.g. all `waiting`, then those `waiting` executions have not made it to the Bull queue yet, so there is nothing to reconcile. The requirements in the [original story](https://linear.app/n8n/issue/PAY-1540) mentioned only `new` and `running`. 

Please let me know if this is what you meant.